snowdrop-execution/src/Snowdrop/Execution/DbActions/AVLp/Actions.hs

```diff
@@ -16,72 +17,73 @@ import qualified Data.Tree.AVL as AVL
-                                                          clientModeToTempSt, reThrowAVLEx,
+                                                          AVLServerState (..), AllAvlEntries,
+                                                          AvlProof (..), AvlProofs, ClientTempState,
+                                                          IsAvlEntry, RetrieveF, RetrieveImpl,
+                                                          RootHashComp (..), clientModeToTempSt,
                                                           runAVLCacheT)
 import           Snowdrop.Execution.DbActions.Types (ClientMode (..), DbAccessActions (..),
                                                      DbModifyActions (..), RememberForProof (..))
-
+import           Snowdrop.Util
 
 avlClientDbActions
-    :: forall k v m h n.
-    ( KVConstraint k v
-    , MonadIO m
-    , MonadCatch m
+    :: forall h xs m n.
+    ( MonadIO m, MonadCatch m, MonadIO n, MonadCatch n
     , AvlHashable h
-    , AVL.Hash h k v
-    , MonadIO n
-    , MonadCatch n
-    , RetrieveImpl (ReaderT (ClientTempState h k v n) m) h
-    , Serialisable (MapLayer h k v h)
+    , RetrieveImpl (ReaderT (ClientTempState h xs n) m) h
+    , AllAvlEntries h xs
     )
     => RetrieveF h n
-    -> RootHash h
-    -> n (ClientMode (AVL.Proof h k v) -> DbModifyActions (AVLChgAccum h k v) k v m ())
+    -> RootHashes h xs
+    -> n (ClientMode (AvlProofs h xs) -> DbModifyActions (AVLChgAccums h xs) xs m ())
 avlClientDbActions retrieveF = fmap mkActions . newTVarIO
   where
     mkActions
-        :: TVar (RootHash h)
-        -> ClientMode (AVL.Proof h k v)
-        -> DbModifyActions (AVLChgAccum h k v) k v m ()
+        :: TVar (RootHashes h xs)
+        -> ClientMode (AvlProofs h xs)
+        -> DbModifyActions (AVLChgAccums h xs) xs m ()
     mkActions var ctMode =
-        DbModifyActions (mkAccessActions var ctMode) (reThrowAVLEx @k . apply var)
+        DbModifyActions (mkAccessActions var ctMode) (apply var)
 
     mkAccessActions
-        :: TVar (RootHash h)
-        -> ClientMode (AVL.Proof h k v)
-        -> DbAccessActions (AVLChgAccum h k v) k v m
-    mkAccessActions var ctMode =
+        :: TVar (RootHashes h xs)
+        -> ClientMode (AvlProofs h xs)
+        -> DbAccessActions (AVLChgAccums h xs) xs  m
+    mkAccessActions var ctMode = do
         DbAccessActions
           -- adding keys to amsRequested
-          (\cA req -> reThrowAVLEx @k $ query cA req =<< createState)
-          (\cA cs -> reThrowAVLEx @k $ modAccum cA cs =<< createState)
+          (\cA req -> createState >>= \ctx -> query ctx cA req)
+          (\cA cs -> createState >>= \ctx -> modAccum ctx cA cs)
           -- setting amsRequested to AMSWholeTree as iteration with
           -- current implementation requires whole tree traversal
-          (\cA p b f -> reThrowAVLEx @k $ iter cA p b f =<< createState)
+          (\cA -> createState >>= \ctx -> iter ctx cA)
```

Comment:
Function `createState` may also throw an exception, thus must be put under `reThrowAVLEx`